### PR TITLE
[Application] Support pre-building reverse proxy image

### PR DESCRIPTION
**Pull request by TomerShor**
Add an option to pre-deploy the reverse proxy image for using it in application runtime.

The 2 main reasons for this are:
1. Speed up deployment - if specifying an existing application image, the build time is only for the reverse proxy. If it is pre-built the function can be deploying much quicker.
2. Using application runtime in air-gapped environments - There is a known issue where golang images cannot be built in an air-gapped environment. To overcome this, users can build the reverse proxy image (golang image) in an exposed environment, and save it in the air-gapped docker registry. This can later be used when creation application functions.

### Usage:
```python
import mlrun

# build the reverse proxy image
mlrun.runtimes.nuclio.application.ApplicationRuntime.deploy_reverse_proxy_image()

# the image is saved on the class:
mlrun.runtimes.nuclio.application.ApplicationRuntime.reverse_proxy_image
```

When the image is set on the class, it will automatically be used by following deployments of instances.

For air-gapped systems, we can also use:
```python
application.from_image(reverse_proxy_image)
```

https://iguazio.atlassian.net/browse/ML-7236

### CODE REVIEW

File: mlrun/runtimes/nuclio/application/application.py [hunk 1]
```diff
@@ -488,6 +489,39 @@ def invoke(
...
+        """
+        Build the reverse proxy image and save it.
+        The reverse proxy image is used to route requests to the application sidecar.
+        This is useful when you want to decrease build time by building the application image only once.
+
+        :param use_cache:   Use the cache when building the image
+        """
+        # create a function that includes only the reverse proxy, without the application
+        from mlrun import get_run_db
+        from mlrun.run import new_function
```
Comment (moranbental) [inline mlrun/runtimes/nuclio/application/application.py]: Why is this import being added inside the function instead of at the top of the file? 
Also, use `import X` instead of `from X import Y`

File: mlrun/runtimes/nuclio/application/application.py [hunk 2]
```diff
@@ -527,21 +561,22 @@ def _build_application_image(
             with_mlrun=with_mlrun,
         )
 
-    def _ensure_reverse_proxy_configurations(self):
-        if self.spec.build.functionSourceCode or self.status.container_image:
+    @classmethod
+    def _ensure_reverse_proxy_configurations(cls, function):
```
Comment (moranbental) [inline mlrun/runtimes/nuclio/application/application.py]: I think this method should be a static method rather than a class method because it does not change any class state or use any class data. 
Also, `get_filename_and_handler` should also be a static method for the same reason.

**Review by moranbental — CHANGES_REQUESTED** [review]

File: mlrun/runtimes/nuclio/application/application.py (hunk 1 above)
Comment (TomerShor) [reply mlrun/runtimes/nuclio/application/application.py]: no actual reason, just copy-paste thing.

**Review by moranbental — APPROVED** [review]
LGTM

File: mlrun/runtimes/nuclio/application/application.py [hunk 3]
```diff
@@ -527,21 +562,22 @@ def _build_application_image(
             with_mlrun=with_mlrun,
         )
 
-    def _ensure_reverse_proxy_configurations(self):
-        if self.spec.build.functionSourceCode or self.status.container_image:
+    @staticmethod
+    def _ensure_reverse_proxy_configurations(function):
+        if function.spec.build.functionSourceCode or function.status.container_image:
```
Comment (rokatyy) [inline mlrun/runtimes/nuclio/application/application.py]: What is the reason of making this method static if it changes the `self`?

Comment (TomerShor) [reply mlrun/runtimes/nuclio/application/application.py]: I wanted to reuse this in the new class method `deploy_reverse_proxy_image`, where we actually create a `RemoteRuntime` instance. to ensure this is done on this instance, I changed the function to get a function instance as an argument, then it can be a non-application runtime function.
Moving this method to the base `RemoteRuntime` didn't feel right, because it is still related to application runtime.

Comment (rokatyy) [reply mlrun/runtimes/nuclio/application/application.py]: @TomerShor I see. `self._ensure_reverse_proxy_configurations(self)` looks a bit weird, so maybe worth adding a comment explaining the trick. Also, please add type hint for a `function` arg

File: server/api/launcher.py [hunk 4]
```diff
@@ -280,6 +280,7 @@ def enrich_runtime(
             not runtime.spec.image
             and not runtime.requires_build()
             and runtime.kind in mlrun.mlconf.function_defaults.image_by_kind.to_dict()
+            and not runtime.skip_base_image_enrichment()
```
Comment (alonmr) [inline server/api/launcher.py]: the enrichment is for `spec.image` but method says base image its not accurate

File: mlrun/runtimes/nuclio/function.py [hunk 5]
```diff
@@ -707,6 +707,12 @@ def enable_default_http_trigger(
         """
         self.spec.disable_default_http_trigger = False
 
+    def skip_base_image_enrichment(self):
+        # make sure the API does not enrich the base image with mlrun/mlrun if the function is not a python function
```
Comment (alonmr) [inline mlrun/runtimes/nuclio/function.py]: `mlrun/mlrun` is magic. what if we decide to make a dedicated nuclio image?

File: mlrun/runtimes/nuclio/application/application.py [hunk 6]
```diff
@@ -488,6 +491,39 @@ def invoke(
...
+        reverse_proxy_func = mlrun.run.new_function(
+            name="reverse-proxy-temp", kind="remote"
+        )
+        # default max replicas is 4, we only need one replica for the reverse proxy
+        reverse_proxy_func.spec.max_replicas = 1
+
+        # clear out the base image, so it won't be mlrun/mlrun
+        reverse_proxy_func.set_config("spec.build.baseImage", None)
+        reverse_proxy_func.spec.image = ""
+        reverse_proxy_func.spec.build.base_image = ""
```
Comment (alonmr) [inline mlrun/runtimes/nuclio/application/application.py]: why is this needed as well as `skip_base_image_enrichment` ?
a more fitting comment would be "the reverse proxy image should not be based on another image" (or use whatever magic nuclio does

File: mlrun/runtimes/nuclio/function.py [hunk 7]
```diff
@@ -707,6 +707,12 @@ def enable_default_http_trigger(
...
+        return (
+            self.spec.nuclio_runtime is None or "python" not in self.spec.nuclio_runtime
```
Comment (alonmr) [inline mlrun/runtimes/nuclio/function.py]: when will nuclio_runtime be None?

File: server/api/launcher.py (hunk 4 above)
Comment (TomerShor) [reply server/api/launcher.py]: Right, its only because Nuclio refers to `spec.image` as the base image.
I'll change the function name

File: mlrun/runtimes/nuclio/application/application.py (hunk 6 above)
Comment (TomerShor) [reply mlrun/runtimes/nuclio/application/application.py]: Because of this: https://github.com/mlrun/mlrun/blob/2bd77151e1c1a1051ddfd141cecc191cc79dbcc2/server/api/crud/runtimes/nuclio/function.py#L686-L690

Updated the comment

File: mlrun/runtimes/nuclio/function.py (hunk 7 above)
Comment (TomerShor) [reply mlrun/runtimes/nuclio/function.py]: In most cases actually.
If it is not `python`, nuclio-jupyter enriches this field.
If it is python, we resolve the runtime in the server from the config:
https://github.com/mlrun/mlrun/blob/2bd77151e1c1a1051ddfd141cecc191cc79dbcc2/server/api/crud/runtimes/nuclio/function.py#L437-L442

File: mlrun/runtimes/nuclio/application/application.py (hunk 6 above)
Comment (alonmr) [reply mlrun/runtimes/nuclio/application/application.py]: yes but its just sanity right? because these should be empty anyway

File: mlrun/runtimes/nuclio/function.py (hunk 7 above)
Comment (alonmr) [reply mlrun/runtimes/nuclio/function.py]: so if it is None and then in the server we resolve it to python. the image should be enriched no?

File: mlrun/runtimes/nuclio/application/application.py (hunk 6 above)
Comment (TomerShor) [reply mlrun/runtimes/nuclio/application/application.py]: Generally the should be empty but nothing stops the user from setting them.
So yes, sanity.

File: mlrun/runtimes/nuclio/function.py (hunk 7 above)
Comment (TomerShor) [reply mlrun/runtimes/nuclio/function.py]: Right I see it now.
I'll change it to:
```python
return self.spec.nuclio_runtime and "python" not in self.spec.nuclio_runtime
```

**Review by alonmr — APPROVED** [review]
Nice! 👍
